[FX] Implement wrap() by patching module globals during symtrace

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50184 [FX] Make len traceable and scriptable with wrap
* **#50182 [FX] Implement wrap() by patching module globals during symtrace**

Differential Revision: [D25819730](https://our.internmc.facebook.com/intern/diff/D25819730)

Commit:
[FX} Implement wrap() by patching module globals during symtrace

[ghstack-poisoned]

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -238,6 +238,60 @@ def proxy_placeholder(name: str):
 
         return root_fn, args
 
+    def call_wrapped_func(self, orig_fn, args, kwargs):
```
re the discussion about this being a class method of tracer: if you want to override the behavior of how a wrapped function behaves, why isn't overriding `create_proxy` and doing something there equivalent? I feel like it would be better to handle all places were we would normally insert a call_function node the same rather than making wrapped functions behave differently than functions override by `__torch_function__`. In the future we may chose to figure out how to wrap execution differently (such as using python tracing to do it), and allowing different behavior based on how an implementation detail of a function is patched would be hard to adapt.

jamesr66a replied:
@mrzzd

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -238,6 +238,60 @@ def proxy_placeholder(name: str):
+        """
+        orig_fns : Dict['frame', Dict['str', Any]] = {}
+
+        global _wrapped_fn_patch_table
```
nit: if you are not assigning it, no need for `global`

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -307,13 +361,52 @@ def forward(*args, **kwargs):
+        raise NotImplementedError('wrap must be called at the top level of a module')
+
+    global _wrapped_fn_patch_table
+    _wrapped_fn_patch_table.setdefault(f, {})
```
It's not a good idea to capture the frame here, it holds `f_back` references to the entire import tree of the file and will probably cause a bunch of objects to stay reachable. Instead, we probably just want to record something like tuples of `(global_dictionary, symbol)`

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -238,6 +238,60 @@ def proxy_placeholder(name: str):
+        else:
+            return orig_fn(*args, **kwargs)
+
+    def _patch_wrapped_functions(self) -> Dict['frame', Dict['str', Any]]:
```
probably not a method on tracer since self is unused.

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -238,6 +238,60 @@ def proxy_placeholder(name: str):
+
+        return orig_fns
+
+    def _unpatch_wrapped_functions(self, orig_fns : Dict['frame', Dict['str', Any]]):
```
self is also unused here

zdevito commented on test/test_fx.py:
```diff
@@ -37,6 +37,11 @@ def forward(self, x):
+def a_lifted_leaf(a, b):
+    return a[0] + a[1] + b
+
+wrap('a_lifted_leaf')
```
Note: we can also allow @wrap on functions to still work. In the use where the argument to wrap is a function rather than a string, simple get the name from `__code__.co_name`. Then you can just write it as an annotation as before.

zdevito reviewed (commented):
Looks good, only some minor stuff about leaking frames.

Commit:
Update on "[FX] Implement wrap() by patching module globals during symtrace"


Differential Revision: [D25819730](https://our.internmc.facebook.com/intern/diff/D25819730)

[ghstack-poisoned]

Commit:
Update on "[FX] Implement wrap() by patching module globals during symtrace"


Differential Revision: [D25819730](https://our.internmc.facebook.com/intern/diff/D25819730)

[ghstack-poisoned]

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -307,13 +307,140 @@ def forward(*args, **kwargs):
             torch.nn.Module.__getattr__ = orig_getattr  # type: ignore
         return self.graph
 
+class _GlobalDictWrapper:
```
I don't think it is necessary to hash the global dictionaries. The patch table can just be:
```
_wrapped_fn_patch_list : List[Tuple[dict, str]]
```

Grouping the patches by global dictionary really doesn't save any time anywhere. True you deduplicate things if someone wrapped something twice but that shouldn't actually happen in a way where it would seriously affect performance. Wrapping an already wrapped function would just be a no-op.

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -307,13 +307,140 @@ def forward(*args, **kwargs):
+    the listed global functions in the `call_wrapped_func` wrapper. Returns
+    a map from frame to dict mapping name to the original global function.
+    """
+    orig_fns : Dict[_GlobalDictWrapper, Dict['str', Any]] = {}
```
This can just be `List[callable]` of the original functions in the same order as the wrap list.

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -307,13 +307,140 @@ def forward(*args, **kwargs):
+        fn_name (Union[str, Callable]): The function or name of the global function to insert into the
+            graph when it's called
+    """
+    if callable(fn_or_name):
```
safer to check for `.__code__` directly. Not all callables are code (e.g. `__list__`)

zdevito reviewed (commented):
Couple of nits.

Commit:
Update on "[FX] Implement wrap() by patching module globals during symtrace"


Differential Revision: [D25819730](https://our.internmc.facebook.com/intern/diff/D25819730)

[ghstack-poisoned]

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -307,13 +307,136 @@ def forward(*args, **kwargs):
+        orig_fn = frame_dict[name]
+        orig_fns.append(PatchedFn(frame_dict, name, orig_fn))
+
+        def scope(orig_fn):
```
If it was `_create_wrapped_func` rather than `_call_wrapped_func` this scoping hack would not be needed.

Commit:
Update on "[FX] Implement wrap() by patching module globals during symtrace"


Differential Revision: [D25819730](https://our.internmc.facebook.com/intern/diff/D25819730)

[ghstack-poisoned]
